src/impl.ml

```diff
@@ -879,17 +879,36 @@ let serve_chunked_to_raw _ c dest _ _ _ _ =
     end in
   loop ()
 
+(* If we're using unbuffered IO, we write in whole sectors. We therefore might
+  need to extend the cstruct to the next sector boundary *)
+let round_up_to_sector unbuffered len =
+  if unbuffered then
+    let sector_size = 512 in
+    (((len - 1) / sector_size) + 1) * sector_size
+  else
+    len
+
 let serve_raw_to_raw common size c dest _ progress _ _ =
   let twomib = 2 * 1024 * 1024 in
   let buffer = IO.alloc twomib in
   let p = progress size in
   let rec loop offset remaining =
-    let this = Int64.(to_int (min remaining (of_int (Cstruct.len buffer)))) in
-    let block = Cstruct.sub buffer 0 this in
-    c.Channels.really_read block >>= fun () ->
+    let n = Int64.(to_int (min remaining (of_int (Cstruct.len buffer)))) in
```

Comment:
Looks like if `size = 0` (and thus `remaining = 0`) we might get a problem when we round up here. I think we should check and rule that out